rust/src/detect/vlan_id.rs

```diff
@@ -0,0 +1,161 @@
+/* Copyright (C) 2024 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+use super::uint::{detect_parse_uint, DetectUintData};
+use std::ffi::CStr;
+use std::str::FromStr;
+
+#[repr(C)]
+#[derive(Debug, PartialEq)]
+pub struct DetectVlanIdData {
+    pub du16: DetectUintData<u16>,
+    pub layer: i8,
+}
+
+pub fn detect_parse_vlan_id(s: &str) -> Option<DetectVlanIdData> {
+    let parts: Vec<&str> = s.split(',').collect();
+    let du16 = detect_parse_uint(parts[0]);
+    if du16.is_err() {
+        return None;
+    }
+    let du16 = du16.unwrap().1;
```

Comment:
@jasonish do you have proposals for better rust style here ?

Reply:
I probably would have done something like:
```rust
let du16 = if let Ok(du16) = detect_parse_uint::<u16>(parts[0]) {
    du16
} else {
    return None;
};
```
Even tho it's not really shorter, but does remove the unwraps.

But just today I learned that you can use `?` on `Option` return types which lets us replace:
```rust
    let du16 = detect_parse_uint(parts[0]);
    if du16.is_err() {
        return None;
    }
    let du16 = du16.unwrap().1;
```
with
```
let du16 = detect_parse_uint(parts[0]).ok()?.1;
```